### PR TITLE
Add Optuna Example for Skorch

### LINKED ISSUE #718: Add an optuna example
It would be nice to add an example of using [optuna](https://github.com/optuna/optuna) with skorch. The main reason is that `GridSearchCV` and `RandomizedSearchCV` can be much too slow for certain neural networks. Optuna has the potential to be much faster.

[This example](https://github.com/optuna/optuna/blob/master/examples/pytorch_simple.py) could be a starting point. As to how, possibilities are a Python script, a Jupyter notebook, or an extension of the docs. Maybe even some of those, like adding a code example _and_ a new chapter in the docs about hyper parameter search.

### PR DESCRIPTION
**Pull request by ParagEkbote**
Fixes #718. 

As previously discussed, I have taken the example from the optuna-integration repo and fixed a couple of errors to ensure that the example runs correctly.

Output:

Number of finished trials: 100
Best trial:
  Value: 0.9507142857142857
  Params: 
    n_layers: 1
    dropout: 0.36848832612373067
    n_units_l0: 104

Could you please review?

cc: @BenjaminBossan

### CODE REVIEW

File: examples/optuna/skorch_example.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+y = y[indices][:N]
+
+X /= 255.0
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+class ClassifierModule(nn.Module):
+    def __init__(self, trial: optuna.Trial) -> None:
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:36]: Let's change the API a bit. Instead of passing the `trial` instance, let's pass `n_layers`, `dropout`, and `output_dim` directly. Inside of `objective`, we can get a suggestion from the `trial` instance and then initialize the model with these hyper-parameters.

File: examples/optuna/skorch_example.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            input_dim = output_dim
+
+        layers.append(nn.Linear(input_dim, 10))
+
+        self.model = nn.Sequential(*layers)
+
+    def forward(self, x):
+        if isinstance(x, dict):
+            x = x["data"] 
+        return F.softmax(self.model(x), dim=-1)
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:58]: Instead of calling `F.softmax` here, it's easier to append an `nn.Softmax(dim=-1)` layer to the `layers` list during `__init__`.

File: examples/optuna/skorch_example.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def objective(trial: optuna.Trial) -> float:
+    net = skorch.NeuralNetClassifier(
+        ClassifierModule(trial),
+        max_epochs=20,
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:64]: I think it would be cool if we could also optimize `max_epochs` and/or `lr` with optuna to demonstrate how to optimize hyper parameters that are not directly tied to the module.

File: examples/optuna/skorch_example.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        action="store_true",
+        help="Activate the pruning feature. `MedianPruner` stops unpromising "
+        "trials at the early stages of training.",
+    )
+    args = parser.parse_args()
+
+    pruner = optuna.pruners.MedianPruner() if args.pruning else optuna.pruners.NopPruner()
+
+    study = optuna.create_study(direction="maximize", pruner=pruner)
+    study.optimize(objective, n_trials=100, timeout=600)
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:89]: Could you please make `n_trials` and possibly `timeout` arguments that can be changed via CLI?

File: examples/optuna/skorch_example.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return F.softmax(self.model(x), dim=-1)
+
+
+def objective(trial: optuna.Trial) -> float:
+    net = skorch.NeuralNetClassifier(
+        ClassifierModule(trial),
+        max_epochs=20,
+        lr=0.1,
+        device=device,
+        callbacks=[SkorchPruningCallback(trial, "valid_acc")],
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:67]: I wonder if setting `verbose=0` and `train_split=False` wouldn't make sense here. This is typically also recommended when doing a normal grid search. WDYT?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for working on the optuna example. It generally looks good already but I made some suggestions how it could be improved. Please take a look.

File: CHANGES.md [hunk 6]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+- Add Optuna Example for Skorch()
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply CHANGES.md]: Updated as discussed.

File: examples/optuna/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+# Installation
+
+To install all the dependencies for the example, run:
+
+```bash
+pip install -r requirements.txt
+```
+
+You can run this example as follows, pruning can be turned on and off with the `--pruning`
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/README.md]: Updated as discussed.

File: examples/optuna/README.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# Optimize Multi-Layer Perceptrons using Optuna and Skorch
+
+## Description
+
+This example is adapted from the optuna-integration [repo](https://github.com/optuna/optuna-examples/blob/main/pytorch/skorch_simple.py).In this example, we optimize the validation accuracy of hand-written digit recognition using
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/README.md]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 5 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py:73]: I found that if we set train_split=False, the training history is not recorded for hyperparameter optimization. I've added the verbose=0 parameter.

File: examples/optuna/skorch_example.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    )
+
+    net.fit(X_train.to_numpy().astype(np.float32), y_train)
+
+    return accuracy_score(y_test.to_numpy(), net.predict(X_test.to_numpy().astype(np.float32)))
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py:78]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 4 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py:109]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 3 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 1 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+argparse
+urllib3
+numpy
+optuna
```
Comment (BenjaminBossan) [inline examples/optuna/requirements.txt:4]: I actually also needed this package: `optuna-integration[skorch]`

File: examples/optuna/skorch_example.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    print("Number of finished trials: {}".format(len(study.trials)))
+
+    print("Best trial:")
+    trial = study.best_trial
+
+    print("  Value: {}".format(trial.value))
+
+    print("  Params: ")
+    for key, value in trial.params.items():
+        print("    {}: {}".format(key, value))
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:124]: Let's put all of this code into a separate `main` function.

File: examples/optuna/skorch_example.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        type=int,
+        default=600,
+        help="Timeout in seconds for the study (default: 600).",
+    )
+    args = parser.parse_args()
+
+    pruner = optuna.pruners.MedianPruner() if args.pruning else optuna.pruners.NopPruner()
+
+    study = optuna.create_study(direction="maximize", pruner=pruner)
+    study.optimize(objective, n_trials=100, timeout=600)
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:113]: You're not using `args.n_trials` here.

File: examples/optuna/README.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+```bash
+pip install -r requirements.txt
+```
+
+You can run this example as follows, pruning can be turned on with the `--pruning`
+argument.
+```bash
+    $ python skorch_example.py [--pruning]
+```
```
Comment (BenjaminBossan) [inline examples/optuna/README.md:21]: Let's either give a short description of each argument this script takes, or mention that there are more arguments that the user can learn about by calling `python skorch_example.py --help`.

File: examples/optuna/skorch_example.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        return self.model(x)
+
+X_train_np = X_train.to_numpy().astype(np.float32)
+X_test_np = X_test.to_numpy().astype(np.float32)
+y_test_np = y_test.to_numpy()
+
+def objective(trial: optuna.Trial) -> float:
+    # Define hyperparameter search space
+    n_layers = trial.suggest_int("n_layers", 1, 3)
+    dropout = trial.suggest_float("dropout", 0.2, 0.5)
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:65]: Let's allow the value to go to 0.

```suggestion
    dropout = trial.suggest_float("dropout", 0.0, 0.5)
```

File: examples/optuna/skorch_example.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    net = skorch.NeuralNetClassifier(
+        model,
+        max_epochs=trial.suggest_int("max_epochs", 10, 50),
+        lr=trial.suggest_float("lr", 1e-4, 1e-1, log=True),
+        device=device,
+        verbose=0,
+        callbacks=[SkorchPruningCallback(trial, "valid_acc")],
+    )
+
+    net.fit(X_train.to_numpy().astype(np.float32), y_train)
```
Comment (BenjaminBossan) [inline examples/optuna/skorch_example.py:80]:
```suggestion
    net.fit(X_train, y_train)
```

This is already done, right?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the updates. I did another pass and still found a few smaller issues, but nothing major. Please check.

File: CHANGES.md [hunk 16]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+- Add Optuna Example for skorch.(#1098)
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply CHANGES.md]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 14 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 15 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/README.md (hunk 13 above)
Comment (ParagEkbote) [reply examples/optuna/README.md:26]: Updated as discussed.

File: examples/optuna/requirements.txt (hunk 10 above)
Comment (ParagEkbote) [reply examples/optuna/requirements.txt:4]: Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 12 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Thanks for catching that:) Updated as discussed.

File: examples/optuna/skorch_example.py (hunk 11 above)
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+        self.model = nn.Sequential(*layers)
+
+    def forward(self, x):
+        if isinstance(x, dict):
+            x = x["data"]
+        return self.model(x)
+
+X_train_np = X_train.to_numpy().astype(np.float32)
+X_test_np = X_test.to_numpy().astype(np.float32)
+y_test_np = y_test.to_numpy()
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+import pandas as pd
+
+from sklearn.datasets import fetch_openml
+from sklearn.metrics import accuracy_score
+from sklearn.model_selection import train_test_split
+
+
+SUBSET_RATIO = 0.4
+
+mnist = fetch_openml("mnist_784", cache=False)
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

**Comment by ParagEkbote** [comment]
I have modularized the script further as suggested. Could you please review?

File: examples/optuna/skorch_example.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+X, y = load_mnist_data(subset_ratio=0.4)
+
+indices = np.random.permutation(len(X))
+N = int(len(X) * 0.4)  # Directly using the subset ratio instead of undefined SUBSET_RATIO
+X, y = X.iloc[indices][:N], y.iloc[indices][:N]
+
+X /= 255.0
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

File: examples/optuna/skorch_example.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    parser.add_argument(
+        "--timeout",
+        "-t",
+        type=int,
+        default=600,
+        help="Timeout in seconds for the study (default: 600).",
+    )
+    args = parser.parse_args()
+
+def main():
```
[the remark this answers is not in the document]
Comment (ParagEkbote) [reply examples/optuna/skorch_example.py]: Updated as discussed.

**Comment by ParagEkbote** [comment]
The failing test is not related to this example. Could you please let me know?

**Comment by ParagEkbote** [comment]
@BenjaminBossan Thank you for reviewing and guiding with adding the example. I've also corrected the original example in the optuna-examples repo [here](https://github.com/optuna/optuna-examples/pull/303).
